### ISSUE TITLE
Consider virtual desktop integration/features

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#2627]
See this discussion thread:

https://groups.google.com/d/topic/qubes-users/gCklOzk9xYg/discussion

**Comment by evadogstar** [comment]
The issue does not describe the problem and what need to do. After reading linked disscussion thread I found that users want some tool to manage windows on the screen and at the virtual desktops. I also think that it's "must have" option.

**Comment by unman** [comment]
I think Andrew had it right in that thread - *some* users want a tool, and with respect @evadogstar , it's not "must have", not even a "must have option". You, perhaps, must have it, but many people will manage perfectly without it.
In my experience many users wont use virtual desktops and find them confusing. If Qubes were to enforce their use it would, I think,reduce uptake.

**Comment by evadogstar** [comment]
I tried to install `devilspie2` for testing purposes on Qubes. Currently, it's not work on Qubes. 
Devilspie2 function `get_window_name()` return Windows names without AppVM labels. Therefore, it's not possible to sort windows on desktops by AppVM name.

**Comment by jpouellet** [comment]
@evadogstar If you want to make it work, you may need to patch it to inspect the [X props set by Qubes](https://github.com/QubesOS/qubes-gui-daemon/blob/291eb806c741b35ffd269f7ec8ba695281bb18ea/gui-daemon/xside.c#L319-L336).

**Comment by tonsimple** [comment]
@unman 
I tend to think that having ability to restrict a VM to a specific (xfce) desktop would be extremely useful and would nicely extend the ways by which user can semi-reflectively distinguish VMs.

It might not be a ~must~ have feature (I am having >10 active, "window-having" VMs and well, I'm mostly okay) but it would make life of users like me easier.

Right now I spend 2-5 minutes starting up applications and carefully, very carefully sending windows to their "supposed" desktops (and then drag each to it's supposed monitor, yes, I am running a dual-head setup) 

It's not the worst thing in the world, but if I could just, like, set "windows from [LAN-work-VM] only can appear on desktop 2 when application starts, AND can be sent to other desktop, while windows from [router-man-VM] can only appear  at desktop 6 when application starts, AND CAN NOT leave that desktop, ever", that would be absolutely awesome for me (and I suspect for any user who run a lot of different GUI stuff in >10 different VMs)

**Comment by jpouellet** [comment]
How is this intended to work for people with many more VMs than it makes sense to have virtual desktops? I have dozens. Is this case covered, or is it expected that users will have few (say, less than 10 or so) VMs?

**Comment by jpouellet** [comment]
Or is the idea to create & destroy virtual desktops dynamically as VMs are started & stopped?

**Comment by tonsimple** [comment]
@jpouellet 
Well, any solution for "many VM" issue will be limited (I for one would like to have more colors and patterns, too, my peak VM count was closer to 20)

Right now I do the following:

Each window can "belong" (belonging determined by me, at every startup, which is clumsy) to a monitor on a virtual desktop (I have two monitors)

Windows from different VMs that share a color (windows7 vm shares color with my LAN-browsing VM, both yellow) should avoid showing up on same desktop

So I send windows to different desktops in accordance to which VM they belong to manually, avoiding "color collisions" (most of the times I manage that, but at "Peak VM" I had several "red" windows from different VMs sharing a desktop because there was just too-much-stuff)

After that I carefully place windows to different monitors within each desktop (thankfully chromium and firefox seem to remember window relative position, so they tend to spawn in the "correct monitor" most of the time)

**Comment by jpouellet** [comment]
It's not clear to me what what specifically is the intended benefit of / problem solved by this approach.
- Do you view this as an approach to solving https://github.com/QubesOS/qubes-issues/issues/2646?
- Reducing user attentiveness requirement for differentiating VMs with similar names/colors?
- Somehow mitigating focus-stealing attacks (e.g. by only allowing focus to be stolen by window on same desktop)?
- Solving the problem of identifying which VM a "borderless" window belongs to when multiple of the same color are running? (e.g. nm-applet popup might actually belong to DispVM, because the window has no title in which to display [vm-name] prefix)
- Some other issue(s)?

IMO it would help the discussion to have a common understanding of *why* people want this feature.

**Comment by tonsimple** [comment]
@jpouellet 
it is indeed a (partial) solution to  https://github.com/QubesOS/qubes-issues/issues/2646 (I am using manual VM-desktop-monitor allocation precisely to cope with many VMs having many GUI windows right now)

Any solution to managing many VMs would reduce attention-requirement for managing many VMs

And yes, if "vm bound to a specific desktop" (and not allowed to have its windows on other desktops) is prohibited from stealing focus from other desktops, that would mitigate both focus stealing and that other DoS (if VM creates very large windows, but is restricted to messing with a particular desktop and not anything beyond it, the severity of such behavior would be much lower)

So it seems that "most full featured" implementation of this would indeed solve many more problems than just managing VMs, but it would require not only ability to set a "default window spawn desktop" for VM, but also ability to (optionally) prohibit a VM from having its window on any other desktop.

not sure Devilspie2 can do the later, but I pretty much know nothing about how Devilspie2 works.

**Comment by tonsimple** [comment]
@evadogstar 
It appears that one can get devilspie2 "doing what has to be done" in qubes without any patching by using get_class_instance_name() instead of get_window_name() 

https://groups.google.com/forum/#!msg/qubes-users/iIQlge9XFLE/sC3v3sAUBAAJ

That appears sufficient for my immediate ergonomic needs.
I will try out devilspie2 and report back.

**Comment by evadogstar** [comment]
@tonsimple 
Yes. Tested and with `get_class_instance_name()` it works without patching.

**Comment by 0spinboson** [comment]
I just found out that this can be done pretty neatly with KDE Plasma 5.17, which is quite responsive even on my dual 4k monitor setup, in Qubes 4.1. Quite nice to no longer have to organize apps on boot.

**Comment by ben-grande** [comment]
Current state:

- i3: easy to configure this setting, syntax is simple
- KDE: somehow, I think that using the GUI for this makes it a bit more difficult than on i3, but it works. KDE has "Activities", I normally do regex based assignment to activities and works quite well.
- Xfce: hasn't improved, people are still using deprecated devilspie2 or other 3rd party software, with a very weird syntax

I don't think it is up to Qubes OS to develop such thing that a WM/DE is covering already, and, uhm, not easy to have a default setting. Closing as not planned. If you disagree with the closing of this issue, please state reason.
